File: ai-cookbook-java-backend/src/main/java/in/ai/chatbot/config/service/RagService.java
package in.ai.chatbot.config.service;

import in.ai.chatbot.config.config.RagProperties;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.document.Document;
import org.springframework.ai.vectorstore.SearchRequest;
import org.springframework.ai.vectorstore.VectorStore;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
public class RagService {

    private static final String SOFT_PROMPT = """
            You are a helpful assistant.

            Relevant context from uploaded documents:
            {context}

            Instructions:
            1. If the context contains relevant information, prioritize using it and cite the source when helpful.
            2. If the context doesn't fully cover the query, supplement with your general knowledge and clarify which parts come from documents vs. general knowledge.
            3. Never contradict information in the provided documents.
            4. If uncertain, acknowledge the limitation rather than guessing.
            """;

    private static final String STRICT_PROMPT = """
            You are a helpful assistant that answers strictly from the provided document context.

            Relevant context from uploaded documents:
            {context}

            Instructions:
            1. Answer ONLY based on the context provided above.
            2. If the context contains enough information to answer the question, provide a clear answer with relevant quotes if appropriate.
            3. If the context does not contain sufficient information, reply: "I don't have information about this in the uploaded documents."
            4. Do not use general knowledge or make assumptions beyond what is explicitly in the documents.
            """;

    private final VectorStore vectorStore;
    private final RagProperties props;

    public RagService(VectorStore vectorStore, RagProperties props) {
        this.vectorStore = vectorStore;
        this.props = props;
    }

    public RagContext buildRagContext(String userMessage) {
        return buildRagContext(userMessage, props.getTopK(), props.getSimilarityThreshold(), props.getMode());
    }

    public RagContext buildRagContext(String userMessage, int topK, double similarityThreshold, String mode) {

        List<Document> hits = vectorStore.similaritySearch(
                SearchRequest.builder()
                        .query(userMessage)
                        .topK(topK)
                        .similarityThreshold(similarityThreshold)
                        .build()
        );
        log.debug("RAG [{}] {} chunk(s) found for: '{}'", mode, hits.size(), userMessage);

        boolean strict = "strict".equalsIgnoreCase(mode);

        if (hits.isEmpty()) {
            return strict
                    ? RagContext.shortCircuit("I don't have information about this in the uploaded documents.")
                    : RagContext.noSystemPrompt();
        }

        List<String> sources = hits.stream()
                .map(doc -> doc.getMetadata().getOrDefault("filename", "unknown").toString())
                .distinct()
                .toList();

        String context = hits.stream()
                .map(Document::getText)
                .collect(Collectors.joining("\n\n---\n\n"));

        String prompt = (strict ? STRICT_PROMPT : SOFT_PROMPT).replace("{context}", context);
        log.debug("RAG [{}] {}", mode, prompt);
        return RagContext.withPrompt(prompt, sources);
    }

    /**
     * Verify VectorStore content by running a test similarity search.
     * Returns the count of documents found for a given query.
     */
    public int verifyVectorStoreContent(String testQuery) {
        try {

            log.debug("Verifying vector store content: {}", testQuery);
            List<Document> hits = vectorStore.similaritySearch(
                    SearchRequest.builder()
                            .query(testQuery)
                            .topK(10)
                            .similarityThreshold(0.0)  // Use 0.0 for verification to see all potential matches
                            .build()
            );
            log.info("VectorStore verification query '{}' returned {} documents", testQuery, hits.size());
              return hits.size();
        } catch (Exception e) {
            log.error("VectorStore verification failed: {}", e.getMessage());
            return -1;
        }
    }

    /**
     * List all documents in VectorStore by querying with a generic search.
     * Returns document metadata including filenames and content preview.
     */
    public List<VectorStoreInfo> listVectorStoreDocuments() {
        try {
            List<Document> docs = vectorStore.similaritySearch(
                    SearchRequest.builder()
                            .query("*") // Wildcard to get entries
                            .topK(1000)
                            .similarityThreshold(0.0)
                            .build()
            );
            return docs.stream()
                    .map(doc -> {
                        String text = doc.getText() != null ? doc.getText() : "";
                        return new VectorStoreInfo(
                                doc.getMetadata().getOrDefault("filename", "unknown").toString(),
                                text.length(),
                                text.substring(0, Math.min(100, text.length()))
                        );
                    })
                    .distinct()
                    .toList();
        } catch (Exception e) {
            log.error("Failed to list VectorStore documents: {}", e.getMessage());
            return List.of();
        }
    }

    public List<SearchResult> searchDocuments(String query, int topK, double similarityThreshold) {
        try {
            List<Document> docs = vectorStore.similaritySearch(
                    SearchRequest.builder()
                            .query(query)
                            .topK(topK)
                            .similarityThreshold(similarityThreshold)
                            .build()
            );
            return docs.stream()
                    .map(doc -> {
                        String filename = doc.getMetadata().getOrDefault("filename", "unknown").toString();
                        Object raw = doc.getMetadata().get("distance");
                        double similarity = raw != null ? Math.round((1.0 - Double.parseDouble(raw.toString())) * 1000.0) / 1000.0 : 0.0;
                        String text = doc.getText() != null ? doc.getText() : "";
                        String preview = text.substring(0, Math.min(200, text.length()));
                        return new SearchResult(filename, similarity, preview);
                    })
                    .toList();
        } catch (Exception e) {
            log.error("Document search failed: {}", e.getMessage());
            return List.of();
        }
    }

    public record RagContext(String systemPrompt, boolean shortCircuit, String shortCircuitMessage, List<String> sources) {
        static RagContext withPrompt(String systemPrompt, List<String> sources) {
            return new RagContext(systemPrompt, false, null, sources);
        }
        static RagContext noSystemPrompt() {
            return new RagContext(null, false, null, List.of());
        }
        static RagContext shortCircuit(String message) {
            return new RagContext(null, true, message, List.of());
        }
    }

    public record VectorStoreInfo(String filename, int contentLength, String contentPreview) {}

    public record SearchResult(String filename, double similarity, String contentPreview) {}
}

File: backend/app/services/rag_service.py
from dataclasses import dataclass

from sqlalchemy.orm import Session

from app.config import Settings
from app.services.embedding_service import EmbeddingService
from app.vector.vector_store_adapter import VectorStoreAdapter


@dataclass
class SearchResult:
    """Public document search result with derived similarity score."""

    filename: str
    similarity: float
    content_preview: str


@dataclass
class RagContext:
    """Prompt context and control flags produced by RAG retrieval."""

    system_prompt: str | None
    short_circuit: bool
    short_circuit_message: str | None
    sources: list[str]


class RagService:
    """Retrieve document chunks and assemble strict or soft RAG context."""

    def __init__(self, session: Session, settings: Settings) -> None:
        """Create a RAG service backed by one database session and settings object."""
        self.settings = settings
        self.embeddings = EmbeddingService(settings)
        self.vector_store = VectorStoreAdapter(session)

    def search_documents(
        self,
        query: str,
        top_k: int | None = None,
        similarity_threshold: float | None = None,
    ) -> list[SearchResult]:
        """Search document vectors and convert distance values to similarities."""
        top_k = top_k if top_k is not None else self.settings.rag_top_k
        similarity_threshold = (
            similarity_threshold
            if similarity_threshold is not None
            else self.settings.rag_similarity_threshold
        )

        embedding = self.embeddings.embed(query)
        max_distance = self.vector_store.max_distance_from_similarity_threshold(
            similarity_threshold
        )
        hits = self.vector_store.search_document_vectors(
            embedding, top_k=top_k, max_distance=max_distance
        )

        return [
            SearchResult(
                filename=str(hit.metadata.get("filename", "unknown")),
                similarity=hit.similarity,
                content_preview=(hit.content or "")[:200],
            )
            for hit in hits
        ]

    def build_rag_context(
        self,
        user_message: str,
        top_k: int | None = None,
        similarity_threshold: float | None = None,
        mode: str | None = None,
    ) -> RagContext:
        """Build a prompt from search results, including strict-mode short circuits."""
        mode = (mode or self.settings.rag_mode).lower()
        results = self.search_documents(
            query=user_message,
            top_k=top_k,
            similarity_threshold=similarity_threshold,
        )

        if not results and mode == "strict":
            return RagContext(
                system_prompt=None,
                short_circuit=True,
                short_circuit_message=(
                    "No relevant information found in indexed documents for strict mode."
                ),
                sources=[],
            )

        if not results and mode != "strict":
            return RagContext(
                system_prompt=None,
                short_circuit=False,
                short_circuit_message=None,
                sources=[],
            )

        sources = sorted({result.filename for result in results})
        context_chunks = [result.content_preview for result in results if result.content_preview]
        context_text = "\n\n---\n\n".join(context_chunks)

        if mode == "strict":
            prompt = (
                "You must answer ONLY using the provided document context. "
                "If context does not contain the answer, say so clearly.\n\n"
                f"Context:\n{context_text}"
            )
        else:
            prompt = (
                "Use document context first. If insufficient, you may provide a "
                "best-effort answer.\n\n"
                f"Context:\n{context_text}"
            )

        return RagContext(
            system_prompt=prompt,
            short_circuit=False,
            short_circuit_message=None,
            sources=sources,
        )
